messages-cosmos.proto

```diff
@@ -0,0 +1,52 @@
+syntax = "proto2";
+
+// Sugar for easier handling in Java
+option java_package = "com.keepkey.deviceprotocol";
+option java_outer_classname = "KeepKeyMessageCosmos";
+
+import "types.proto";
+
+/**
+ * Request: Address at the specified index
+ * @start
+ * @next CosmosAddress
+ */
+message CosmosGetAddress {
+    repeated uint32 address_n = 1;  // BIP-32 path. For compatibility with other wallets, must be m/44'/148'/index'
+    optional bool show_display = 2; // optionally show on display before sending the result
+}
+
+/**
+ * Response: Address for the given index
+ * @end
+ */
+message CosmosAddress {
+    optional string address = 1;    // Address in Stellar format (base32 of a pubkey with checksum)
+}
+
+/**
+ * Request: ask device to sign Cosmos transaction
+ * @start
+ * @next CosmosSignedTx
+ */
+message CosmosSignTx {
+    repeated uint32 address_n = 1;          // BIP-32 path. For compatibility with other wallets, must be m/44'/118'/index'
+    required uint64 account_number = 2;
+    required string chain_id = 3;
+    required uint32 fee_amount = 4;
+    required uint32 gas = 5;
+    required string from_address = 6;
+    required string to_address = 7;
+    required uint64 amount = 8;
+    required string memo = 9;
+    required uint64 sequence = 10;
+}
```

Comment:
In the other commit, you've got:

https://github.com/keepkey/device-protocol/compare/88e29410753b26d27882ed66460e8bd61d6980c6...2e297ce5c9550ee72015f9f607359bae15266113#diff-59a453ca48d4fcab5089a5bcfea9247fR37

... which I like. But let's take that a step further:

```
message CosmosSignTx {
    repeated uint32 address_n = 1;
    optional uint64 account_number = 2;
    optional string chain_id = 3; // Is this really a `string`, or should it be `optional bytes chain_id = 3;` ?
    optional CosmosFee fee = 4;
    optional string memo = 5; // We may want to stick this in a CosmosMsgAck, so we can support arbitrary length memos.
    optional uint64 sequence = 6;
    optional uint32 msg_count = 7;
}

message CosmosMsgRequest {
}

message CosmosMsgAck {
    optional CosmosMsgSend send = 1;
    // This leaves room in the protocol for adding more cosmos messages later
}

message CosmosMsgSend {
    optional string from_address = 1; // Do we actually need this, or can we derive it via address_n + user's keys?
    optional string to_address = 2;
    optional uint64 amount = 3;
    // This implicitly assumes we're sending ATOMs. Maybe we should make room for cosmos tokens?
}
```

Sidenote: we should use `optional` everywhere, and do checking in the firmware, since `optional` gives us significantly more upgrade flexibility than `required`.

Reply:
> // We may want to stick this in a CosmosMsgAck, so we can support arbitrary length memos.

We talked about this internally, and I think it's fine to leave the memo as a field on `CosmosSignTx`.

Reply:
am I correct in saying that `CosmosMsgAck` needs to be added to `CosmosSignTx` or am I missing something about how we are going to get the message stream?

Reply:
`*Ack` messages are the host responding to the corresponding `*Request` message that comes from the device. That means, for a two-output transaction, you'd see this on the wire:

```
host -> device: CosmosSignTx { ... msg_count=2 }
device -> host: CosmosMsgRequest {}
host -> device: CosmosMsgAck { send: { /* details for output 1 */ } }
device -> host: CosmosMsgRequest {}
host -> device: CosmosMsgAck { send: { /* details for output 2 */ } }
device -> host: CosmosSignedTx {...}
```

And each time the device gets a `CosmosMsgAck`, it will confirm that output (or whatever it ends up being when we add more cosmos msg types) with the user, and then append the relevant data to the incremental hash. Once it has all the pieces, it returns the signature.

Reply:
This is a bit different than how you've currently architected the protocol, which will mean some changes to the firmware and python bits.... which is why I wanted to chat about the protocol design *before* you got it all implemented.

Reply:
As a reference, you can look at the EOS implementation, which does something similar to allow an arbitrary number of actions in a single tx.

Reply:
Yeah unfortunately the only time we had before the deadline (today) to do this was Sat/Sun, so we couldn't really talk about it prior. I'll take a look at EOS.